### PR TITLE
Add Terminate-if-Running test

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Added a test for Start Workflow + Terminate-if-Running.

Also refactored the existing test cases to make them more explicit, have less duplication, less clever looping.

## Why?

Adding a test for Terminate-if-Running behavior for changing its behavior in an upcoming PR (related to https://github.com/temporalio/api/pull/359). 

## How did you test it?

Is test.

### CODE REVIEW

File: service/history/history_engine2_test.go [hunk 1]
```diff
@@ -1354,104 +1253,230 @@ func (s *engine2Suite) TestStartWorkflowExecution_NotRunning_PrevSuccess() {
...
+			}, nil)
+
+			resp, err := s.historyEngine.StartWorkflowExecution(
+				metrics.AddMetricsContext(context.Background()),
+				makeStartRequest(enumspb.WORKFLOW_ID_REUSE_POLICY_TERMINATE_IF_RUNNING))
+
+			s.NoError(err)
+			s.NotEqual(prevRunID, resp.GetRunId())
+		})
+	})
```
Comment (stephanos) [inline service/history/history_engine2_test.go:1332]: net new test

File: service/history/history_engine2_test.go [hunk 2]
```diff
@@ -1354,104 +1253,228 @@ func (s *engine2Suite) TestStartWorkflowExecution_NotRunning_PrevSuccess() {
...
+			resp, err := s.historyEngine.StartWorkflowExecution(
+				metrics.AddMetricsContext(context.Background()),
+				makeStartRequest(enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE))
+
+			var expectedErr *serviceerror.WorkflowExecutionAlreadyStarted
+			s.ErrorAs(err, &expectedErr)
+			s.Nil(resp)
+		})
+
+		s.Run("terminate when id reuse policy is TERMINATE_IF_RUNNING", func() {
```
Comment (stephanos) [inline service/history/history_engine2_test.go:1302]: net new test

File: service/history/history_engine2_test.go [hunk 3]
```diff
@@ -228,11 +227,19 @@ func (s *engine2Suite) SetupTest() {
 	s.historyEngine = h
 }
 
+func (s *engine2Suite) SetupSubTest() {
+	s.SetupTest()
+}
```
Comment (stephanos) [inline service/history/history_engine2_test.go:232]: This is, unfortunately, necessary to execute the setup before each test _and_ sub test.

File: service/history/history_engine2_test.go [hunk 4]
```diff
@@ -1354,104 +1257,228 @@ func (s *engine2Suite) TestStartWorkflowExecution_NotRunning_PrevSuccess() {
...
+						RequestID:        requestID,
+						RunID:            prevRunID,
+						State:            enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED,
+						Status:           status,
+						LastWriteVersion: lastWriteVersion,
+					}
+				}
+
+				s.Run(fmt.Sprintf("status %v", status), func() {
+					s.Run("and id reuse policy ALLOW_DUPLICATE", func() {
```
Comment (alexshtin) [inline service/history/history_engine2_test.go:1439]: :+1:. You should add a screenshot of test runner with all this nested tests are concatenated to plain English sentence.

**Review by alexshtin — APPROVED** [review]

File: service/history/history_engine2_test.go (hunk 4 above)
Comment (stephanos) [reply service/history/history_engine2_test.go:1439]: Goland only shows them hierarchical:

Screenshot 2024-03-05 at 10 15 42 AM

**Pull request merged**